### PR TITLE
View funcs delegation

### PR DESCRIPTION
**Pull request by sasurobert**
added new view functions to delegation sc
added protection of unStaking / unBonding all tokens when there are active nodes
take out warm instance from config
added integration tests for different scenarios

### CODE REVIEW

File: process/smartContract/process.go [hunk 1]
```diff
@@ -908,7 +909,7 @@ func (sc *scProcessor) isSCExecutionAfterBuiltInFunc(
 			CallValue:      big.NewInt(0),
 			CallType:       callType,
 			GasPrice:       vmInput.GasPrice,
-			GasProvided:    vmOutput.GasRemaining - vmInput.GasLocked,
+			GasProvided:    scExecuteOutTransfer.GasLimit,
```
Comment (iulianpascalau) [inline process/smartContract/process.go:912]: is this backwards compatible?

File: vm/systemSmartContracts/delegation.go [hunk 2]
```diff
@@ -1966,6 +1980,85 @@ func (d *delegation) getClaimableRewards(args *vmcommon.ContractCallInput) vmcom
 	return vmcommon.Ok
 }
 
+func (d *delegation) isDelegator(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/delegation.go:1983]: Please add these functions in the Jira task list so we can have unit tests for them.

File: process/smartContract/process.go (hunk 1 above)
Comment (sasurobert) [reply process/smartContract/process.go:912]: Yes it is. This feature is under built in functions

File: vm/systemSmartContracts/delegation.go (hunk 2 above)
Comment (sasurobert) [reply vm/systemSmartContracts/delegation.go:1983]: Bogdan already did it and opened PR

Comment (iulianpascalau) [reply vm/systemSmartContracts/delegation.go:1983]: We can easily forget about them if the PRs overlap

File: process/smartContract/process.go (hunk 1 above)
Comment (iulianpascalau) [reply process/smartContract/process.go:912]: ok

File: vm/systemSmartContracts/delegation.go (hunk 2 above)
Comment (bogdan-rosianu) [reply vm/systemSmartContracts/delegation.go:1983]: the PR with unit tests is directed to this branch. We are safe

File: vm/systemSmartContracts/delegationManager.go [hunk 3]
```diff
@@ -313,7 +313,7 @@ func (d *delegationManager) getAllContractAddresses(args *vmcommon.ContractCallI
 		return vmcommon.UserError
 	}
 
-	for _, address := range contractList.Addresses {
+	for _, address := range contractList.Addresses[1:] {
```
Comment (BeniaminDrasovean) [inline vm/systemSmartContracts/delegationManager.go:316]: You should add a check that len(contractList.Addresses)>0, otherwise this will panic.

**Review by BeniaminDrasovean — CHANGES_REQUESTED** [review]

File: vm/systemSmartContracts/delegationManager.go (hunk 3 above)
Comment (sasurobert) [reply vm/systemSmartContracts/delegationManager.go:316]: at initialization it will be at least one. I can add for extra security.

**Review by iulianpascalau — CHANGES_REQUESTED** [review]
Function `getUserUnDelegatedList` not tested @bogdan-rosianu. Task EN-8525 is not complete

File: vm/systemSmartContracts/delegation.go (hunk 2 above)
Comment (iulianpascalau) [reply vm/systemSmartContracts/delegation.go:1983]: You forgot to test a function

**Review by BeniaminDrasovean — DISMISSED** [review]

**Review by iulianpascalau — APPROVED** [review]

File: vm/systemSmartContracts/validator.go [hunk 4]
```diff
@@ -1575,6 +1587,11 @@ func (v *validatorSC) unBondTokens(args *vmcommon.ContractCallInput) vmcommon.Re
 		return vmcommon.Ok
 	}
 
+	if registrationData.NumRegistered > 0 && registrationData.TotalStakeValue.Cmp(v.minDeposit) < 0 {
+		v.eei.AddReturnMessage("cannot unStake tokens, the validator would remain without min deposit, nodes are still active")
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/validator.go:1591]: cannot unBond

File: vm/systemSmartContracts/validator_test.go [hunk 5]
```diff
@@ -3492,6 +3534,51 @@ func TestStakingValidatorSC_UnBondTokensShouldWork(t *testing.T) {
...
+					UnstakedValue: big.NewInt(1),
+				},
+			},
+			TotalUnstaked: big.NewInt(10),
+		},
+	)
+
+	callFunctionAndCheckResult(t, "unBondTokens", sc, caller, nil, zero, vmcommon.UserError)
+	vmOutput := eei.CreateVMOutput()
+	assert.True(t, strings.Contains(vmOutput.ReturnMessage, "cannot unStake tokens, the validator would remain without min deposit, nodes are still active"))
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/validator_test.go:3579]: cannot unBond

**Review by iulianpascalau — APPROVED** [review]

**Review by SebastianMarian — DISMISSED** [review]

File: vm/systemSmartContracts/validator.go (hunk 4 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go]: done

File: vm/systemSmartContracts/validator_test.go (hunk 5 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator_test.go]: done

**Review by BeniaminDrasovean — DISMISSED** [review]

File: vm/systemSmartContracts/staking.go [hunk 6]
```diff
@@ -1461,6 +1463,27 @@ func (s *stakingSC) getOwner(args *vmcommon.ContractCallInput) vmcommon.ReturnCo
 	return vmcommon.Ok
 }
 
+func (s *stakingSC) getTotalNumberOfRegisteredNodes(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !s.flagStakingV2.IsSet() {
+		s.eei.AddReturnMessage("invalid method to call")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		return vmcommon.UserError
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/staking.go:1472]: No s.eei.AddReturnMessage ?

File: process/smartContract/builtInFunctions/esdtTransfer.go [hunk 7]
```diff
@@ -120,6 +120,7 @@ func (e *esdtTransfer) ProcessBuiltinFunction(
 		}
 
 		if isSCCallAfter {
+			vmOutput.GasRemaining = vmInput.GasProvided - e.funcGasCost
```
Comment (SebastianMarian) [inline process/smartContract/builtInFunctions/esdtTransfer.go:123]: Could be negative in some edge cases? safeSub ?

File: vm/systemSmartContracts/staking.go (hunk 6 above)
Comment (sasurobert) [reply vm/systemSmartContracts/staking.go:1472]: done

File: process/smartContract/builtInFunctions/esdtTransfer.go (hunk 7 above)
Comment (sasurobert) [reply process/smartContract/builtInFunctions/esdtTransfer.go:123]: I do not think so. but added safeSub

**Review by iulianpascalau — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Review by stefanandy — APPROVED** [review]
System Test Passed

**Pull request merged**